src/CPUActivityAgent.cpp

```diff
@@ -136,46 +138,61 @@ namespace geopm
         m_platform_io.write_control("MSR::QM_EVTSEL:EVENT_ID", GEOPM_DOMAIN_BOARD, 0, 2);
     }
 
-    // Validate incoming policy and configure default policy requests.
-    void CPUActivityAgent::validate_policy(std::vector<double> &in_policy) const
+    void CPUActivityAgent::init_constconfig_io()
     {
-        GEOPM_DEBUG_ASSERT(in_policy.size() == M_NUM_POLICY,
-                           "CPUActivityAgent::" + std::string(__func__) +
-                           "(): policy vector not correctly sized.  Expected  " +
-                           std::to_string(M_NUM_POLICY) + ", actual: " +
-                           std::to_string(in_policy.size()));
+        m_qm_max_rate = {};
+        const auto ALL_NAMES = m_platform_io.signal_names();
 
-        // Check for NAN to set default values for policy
-        if (std::isnan(in_policy[M_POLICY_CPU_FREQ_MAX])) {
-            in_policy[M_POLICY_CPU_FREQ_MAX] = m_freq_core_max;
+        // F efficient values
+        std::string fe_constconfig = "CONST_CONFIG::CPU_FREQUENCY_EFFICIENT_HIGH_INTENSITY";
+        if (ALL_NAMES.count(fe_constconfig) != 0) {
+            m_freq_core_efficient = m_platform_io.read_signal(fe_constconfig, GEOPM_DOMAIN_BOARD, 0);
         }
-
-        // Check for NAN to set default values for policy
-        if (std::isnan(in_policy[M_POLICY_CPU_FREQ_EFFICIENT])) {
-            in_policy[M_POLICY_CPU_FREQ_EFFICIENT] = m_freq_core_min;
+        else {
+            m_freq_core_efficient = m_freq_core_min;
         }
 
-        //////////////////////////
-        //UNCORE POLICY CHECKING//
-        //////////////////////////
-        if (std::isnan(in_policy[M_POLICY_UNCORE_FREQ_MAX])) {
-            in_policy[M_POLICY_UNCORE_FREQ_MAX] = m_freq_uncore_max;
+        fe_constconfig = "CONST_CONFIG::CPU_UNCORE_FREQUENCY_EFFICIENT_HIGH_INTENSITY";
+        if (ALL_NAMES.count(fe_constconfig) != 0) {
+            m_freq_uncore_efficient = m_platform_io.read_signal(fe_constconfig, GEOPM_DOMAIN_BOARD, 0);
         }
-        if (std::isnan(in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT])) {
-            in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = m_freq_uncore_min;
+        else {
+            m_freq_uncore_efficient = m_freq_uncore_min;
         }
 
-        // When the policy is all NaNs, this check also verifies that
-        // the system was not left in a bad state with regard to the
-        // UNCORE_FREQUENCY_<MAX/MIN>_CONTROLs.
-        if (in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] > in_policy[M_POLICY_UNCORE_FREQ_MAX]) {
+        // Grab all (uncore frequency, max memory bandwidth) pairs
+        for (int entry_idx = 0; entry_idx < (int)ALL_NAMES.size(); ++entry_idx) {
```

Comment:
I know we don't follow this strictly in the codebase, but I would recommend using C++-style casts instead of C-style casts.

However, looking at this ``for`` loop more closely, it seems you should be able to use an unsigned int type for ``entry_idx`` and avoid having to do any casts.

Reply:
Yeah, like this:
```cpp
size_t ALL_NAMES_size = ALL_NAMES.size();
for (size_t entry_idx = 0; entry_idx < ALL_NAMES_size; ++entry_idx) {
```